ring/spread_minimizing_token_generator.go

```diff
@@ -0,0 +1,281 @@
+		logger:     logger,
+	}
+	return tokenGenerator, nil
+}
+
+func getInstanceID(instanceID string) (int, error) {
+	parts := instanceIDRegex.FindStringSubmatch(instanceID)
+	if len(parts) != 4 {
+		return -1, errorBadInstanceIDFormat(instanceID)
+	}
+	id, err := strconv.Atoi(parts[3])
+	if err != nil {
+		return -1, err
+	}
+	return id, nil
+}
+
+func getZoneID(zone string, zones []string) (int, error) {
+	if !slices.IsSorted(zones) {
+		slices.Sort(zones)
+	}
+	index := sort.SearchStrings(zones, zone)
+	if index >= len(zones) {
+		return -1, errorZoneNotValid(zone)
+	}
+	return index, nil
+}
+
+// generateInitTokens calculates a set of tokens for a given zone that will be assigned
+// to the first instance (with id 0) of that zone.
+func (t *SpreadMinimizingTokenGenerator) generateInitTokens(zoneID int, tokensPerInstance int) (Tokens, error) {
+	tokenDistance := float64(1<<32) / float64(tokensPerInstance)
+	zonesCount := uint32(len(t.zones))
+	tokenDistanceLen := bits.Len(uint(tokenDistance) - 1)
+	zonesOffsetLen := bits.Len(uint(len(t.zones) - 1))
+	if zonesOffsetLen >= tokenDistanceLen {
+		return nil, errorZoneSetTooBig(len(t.zones), tokensPerInstance)
+	}
+	tokens := make(Tokens, 0, tokensPerInstance)
+	for i := 0; i < tokensPerInstance; i++ {
+		token := uint32(i << tokenDistanceLen)
+		token = (token/zonesCount)*zonesCount + uint32(zoneID)
+		tokens = append(tokens, token)
+	}
+	return tokens, nil
+}
+
+// calculateNewToken determines where in the range represented by the given ringToken should a new token be placed
+// in order to satisfy the constraint represented by the optimalTokenOwnership. This method assumes that:
+// - ringToken.token % zonesCount == ringToken.prevToken % zonesCount
+// - optimalTokenOwnership % zonesCount == 0,
+// where zonesCount is the number of zones in the ring. The caller of this function must ensure that these assumptions hold.
+func (t *SpreadMinimizingTokenGenerator) calculateNewToken(token *ringToken, optimalTokenOwnership uint32) (uint32, error) {
+	zonesCount := uint32(len(t.zones))
+	if optimalTokenOwnership < zonesCount || optimalTokenOwnership%zonesCount != 0 {
+		return 0, errorMultipleOfZonesCount(zonesCount, optimalTokenOwnership, token)
+	}
+	if token.prevToken%zonesCount != token.token%zonesCount {
+		return 0, errorLowerAndUpperBoundModulo(zonesCount, optimalTokenOwnership, token)
+	}
+	ownership := getTokenDistance(token.prevToken, token.token)
+	if ownership <= int(optimalTokenOwnership) {
+		return 0, errorDistanceBetweenTokensNotBigEnough(int(optimalTokenOwnership), ownership, token)
+	}
+	maxTokenValue := (math.MaxUint32/zonesCount - 1) * zonesCount
+	offset := maxTokenValue - token.prevToken
+	if offset < optimalTokenOwnership {
+		newToken := optimalTokenOwnership - offset
+		return newToken, nil
+	}
+	return token.prevToken + optimalTokenOwnership, nil
+}
+
+// getOptimalTokenOwnership calculates the optimal ownership of the remaining currTokensCount tokens of an instance
```

Comment:
I find confusing that `currTokensCount` isn't the _current tokens count_ but the _remaining tokens count_.